google-cloud-storage/lib/google/cloud/storage/file/signer_v4.rb

```diff
@@ -341,11 +350,15 @@ def post_object_ext_url scheme, virtual_hosted_style, bucket_bound_hostname
           end
 
           def generate_signature signing_key, data
-            unless signing_key.respond_to? :sign
-              signing_key = OpenSSL::PKey::RSA.new signing_key
+            if signing_key.is_a? Proc
+              signing_key.call(data).unpack("H*").first.force_encoding "utf-8"
```

Comment:
Similarly, consider

```ruby
signature = signing_key.call data
```

and then lifting the rest of the expression out of the if statement.